distribution/cachedir/manager.go

```diff
@@ -0,0 +1,206 @@
+// deleted. If the delete operation fails and returns an error, the key is
+// still considered released.
+func (m *Manager) DeleteDir(key string) error {
+	m.mu.Lock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager delete")
+	if _, ok := m.inUse[hashedKey]; !ok {
+		m.mu.Unlock()
+		panic(fmt.Errorf("tried to delete a cachedir key (%s) without getting it first", key))
+	}
+	delete(m.inUse, hashedKey)
+	m.mu.Unlock()
+	return m.cleanUpByKeyHash(hashedKey)
+}
+
+func (m *Manager) getAllKeyHashes() map[digest.Digest]struct{} {
+	allKeyHashes := map[digest.Digest]struct{}{}
+	algorithmDirs, err := ioutil.ReadDir(m.cacheRoot)
+	if err != nil {
+		if !strings.Contains(err.Error(), "no such file or directory") {
+			logrus.WithError(err).WithField("cacheRoot", m.cacheRoot).Warn("Failed to list cache root")
+		}
+	}
+	for _, algorithmDir := range algorithmDirs {
+		algorithmDirName := algorithmDir.Name()
+		algorithm := filepath.Base(algorithmDirName)
+		cacheDirs, err := ioutil.ReadDir(filepath.Join(m.cacheRoot, algorithm))
+		if err != nil {
+			logrus.WithError(err).WithField("cacheRoot", m.cacheRoot).WithField("algorithmDir", algorithmDir).Warn("Failed to list cache directory")
+		}
+		for _, cacheDir := range cacheDirs {
+			cacheDirName := cacheDir.Name()
+			keyHash := digest.NewDigestFromHex(algorithm, filepath.Base(cacheDirName))
+
+			// safety check to make sure we don't delete anything that's not a hash
+			// we created
+			err := keyHash.Validate()
+			if err != nil {
+				logrus.WithError(err).WithField("cachedir", cacheDirName).Debug("Ignoring unexpected cache directory")
+				continue
+			}
+
+			allKeyHashes[keyHash] = struct{}{}
+		}
+	}
+	return allKeyHashes
+}
+
+// Usage returns information about how much space the download cache is
+// using. If any directories or files are unreadable, we just log the
+// errors and don't count the files.
+func (m *Manager) Usage() types.DownloadCacheUsage {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	allKeyHashes := m.getAllKeyHashes()
+
+	var total uint64
+	var active uint64
+	var totalBytes uint64
+	var activeBytes uint64
+	for keyHash := range allKeyHashes {
+		size := m.sizeEstimateByKeyHash(keyHash)
+		totalBytes += size
+		total++
+		if _, ok := m.inUse[keyHash]; ok {
+			active++
+			activeBytes += size
+		}
+	}
+
+	return types.DownloadCacheUsage{
+		TotalNumber:  int64(total),
+		ActiveNumber: int64(active),
+		TotalBytes:   int64(totalBytes),
+		ActiveBytes:  int64(activeBytes),
+	}
+}
+
+// CollectGarbage deletes all dirs that are not currently in use (ones that
+// don't have a GetDir without a matching ReleaseDir/DeleteDir call) and
+// returns the amount of space saved in bytes.
+func (m *Manager) CollectGarbage() (uint64, error) {
+	// safety check to make sure we don't try to delete files in / or pwd
+	if m.cacheRoot == "" {
+		return 0, nil
```

Comment:
I don't think so, because an empty cacheRoot is a way of opting out from the functionality.